# 1. Data Extraction

In this notebook we will extract our data and analyse it. For that purpose, we are importing from our library where we define the
```bcrp_dataframe``` dataframe. This function will allows us to use the API interface of the Central Bank of Reserve of Peru (BCRP) to automatically create a pandas dataframe with the necessary codes.

## 1.1 Libraries

We import the necessary libraries, including our own library in the modules file

In [33]:
# Warnings
import warnings
warnings.filterwarnings("ignore")

# Basic Libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import requests
import seaborn as sns
from scipy import stats
from functools import reduce

# Statsmodels
import statsmodels.api as sm
import pmdarima as pmd
from pmdarima.arima import auto_arima
from statsmodels.tsa.api import VAR
from statsmodels.tsa.vector_ar.var_model import VARResults
from statsmodels.tsa.statespace.sarimax import SARIMAX
from statsmodels.tsa.arima.model import ARIMA
from statsmodels.tsa.stattools import adfuller
from statsmodels.tsa.seasonal import STL

# Machine Learning models
from sklearn.ensemble import RandomForestRegressor, RandomForestClassifier
from sklearn.model_selection import train_test_split, GridSearchCV, TimeSeriesSplit
from sklearn.linear_model import Ridge, Lasso, ElasticNet, ElasticNetCV, LinearRegression
from sklearn.linear_model import LinearRegression
from sklearn.metrics import (
    mean_absolute_error,
    mean_squared_error,
    mean_absolute_percentage_error,
    median_absolute_error,
    r2_score,
    precision_score

)

from xgboost import XGBRegressor



In [34]:
# We import our own functions
import sys
sys.path.append('../../..')  # Move two levels up to the project root
from modules.functions import *

## 1.2 Extraction
We define our inputs and apply them the ```bcrp_dataframe``` function in order to obtain the pandas dataframe with the corresponding series

We define the following inputs:

    series     = the code of the series we are going to extract
    start_date = the starting date, when the BCRP starts using the interest rate as a policy measure
    end_date   = December 2019
    freq       = Monthly frequency

### df_1
We can now create the first dataframe with the ```bcrp_dataframe``` function. This dataframe contains out target variable Headline Inflation 

In [35]:
series     = ['PN01271PM']
start_date = '2003-09'
end_date   = '2023-12'
freq       = 'Mensual'

In [36]:
df_1 = bcrp_dataframe( series , start_date , end_date , freq )
df_1.head()

Índice de precios Lima Metropolitana (var% mensual) - IPC
Fecha                                                                
2003-09-01                                           0.558598        
2003-10-01                                           0.049032        
2003-11-01                                           0.167685        
2003-12-01                                           0.563951        
2004-01-01                                           0.537447

In [37]:
df_1 = get_trend(df_1)
df_1.head()

Índice de precios Lima Metropolitana (var% mensual) - IPC
Fecha                                                                
2003-09-01                                           0.518834        
2003-10-01                                           0.492247        
2003-11-01                                           0.465462        
2003-12-01                                           0.438413        
2004-01-01                                           0.411138

### df_2
We create the second dataframe with the ```bcrp_dataframe``` function. This dataframe contains rate variables. We use this variables in levels. It is not necessary to differentiate them.

In [38]:
series     = ['PD04722MM', 'PN01207PM']
start_date = '2003-09'
end_date   = '2023-12'
freq       = 'Mensual'

In [39]:
df_2 = bcrp_dataframe( series , start_date , end_date , freq )
df_2.head()

Tasas de interés del Banco Central de Reserva  - Tasa de Referencia de la Política Monetaria  \
Fecha                                                                                                      
2003-09-01                                               2.75                                              
2003-10-01                                               2.75                                              
2003-11-01                                               2.50                                              
2003-12-01                                               2.50                                              
2004-01-01                                               2.50                                              

            Tipo de cambio - promedio del periodo (S/ por US$) - Interbancario - Promedio  
Fecha                                                                                      
2003-09-01                                           3.480898                              
2003-10-01                                           3.478177                              
2003-11-01                                           3.477635                              
2003-12-01                                           3.471176                              
2004-01-01                                           3.467352

### df_3
We create the third dataframe with the ```bcrp_dataframe``` function. This dataframe contains monetary variables as well as commodities. We differentiate those variables in order to get their monthly % change

In [46]:
series     = ['PN00495MM', 'PN06481IM', 'PN02125PM', 'PN01661XM','PN01662XM','PN01664XM','PN01660XM']
start_date = '2003-09'
end_date   = '2023-12'
freq       = 'Mensual'

In [47]:
df_3 = bcrp_dataframe( series , start_date , end_date , freq )
df_3 = np.log(df_3)
df_3 = df_3.dropna()
df_3.head()

Emisión primaria y multiplicador (millones S/) - Circulante Desestacionalizado  \
Fecha                                                                                        
2003-09-01                                           8.591877                                
2003-10-01                                           8.602051                                
2003-11-01                                           8.613430                                
2003-12-01                                           8.629901                                
2004-01-01                                           8.649204                                

            Liquidez internacional del BCRP - RIN - Reservas Internacionales Netas (millones US$)  \
Fecha                                                                                               
2003-09-01                                           9.185581                                       
2003-10-01                                           9.191108                                       
2003-11-01                                           9.240173                                       
2003-12-01                                           9.229584                                       
2004-01-01                                           9.265177                                       

            Remuneraciones - Remuneración Mínima Vital - Índice Real (base 1994 = 100)  \
Fecha                                                                                    
2003-09-01                                           5.440117                            
2003-10-01                                           5.495508                            
2003-11-01                                           5.493832                            
2003-12-01                                           5.488208                            
2004-01-01                                           5.482848                            

            Cotizaciones de productos (promedio del periodo) - Trigo - EEUU (US$ por toneladas)  \
Fecha                                                                                             
2003-09-01                                           4.884206                                     
2003-10-01                                           4.900658                                     
2003-11-01                                           5.009696                                     
2003-12-01                                           5.046169                                     
2004-01-01                                           5.049709                                     

            Cotizaciones de productos (promedio del periodo) - Maíz - EEUU (US$ por toneladas)  \
Fecha                                                                                            
2003-09-01                                           4.419001                                    
2003-10-01                                           4.390800                                    
2003-11-01                                           4.463461                                    
2003-12-01                                           4.511148                                    
2004-01-01                                           4.571827                                    

            Cotizaciones de productos (promedio del periodo) - Aceite Soya - EEUU (US$ por toneladas)  \
Fecha                                                                                                   
2003-09-01                                           6.238332                                           
2003-10-01                                           6.404890                                           
2003-11-01                                           6.421768                                           
2003-12-01                                           6.471752                                           
200

In [48]:
df = df_1.join(df_2).join(df_3)
df.dropna(inplace=True)
df.head()

Índice de precios Lima Metropolitana (var% mensual) - IPC  \
Fecha                                                                   
2003-09-01                                           0.518834           
2003-10-01                                           0.492247           
2003-11-01                                           0.465462           
2003-12-01                                           0.438413           
2004-01-01                                           0.411138           

            Tasas de interés del Banco Central de Reserva  - Tasa de Referencia de la Política Monetaria  \
Fecha                                                                                                      
2003-09-01                                               2.75                                              
2003-10-01                                               2.75                                              
2003-11-01                                               2.50                                              
2003-12-01                                               2.50                                              
2004-01-01                                               2.50                                              

            Tipo de cambio - promedio del periodo (S/ por US$) - Interbancario - Promedio  \
Fecha                                                                                       
2003-09-01                                           3.480898                               
2003-10-01                                           3.478177                               
2003-11-01                                           3.477635                               
2003-12-01                                           3.471176                               
2004-01-01                                           3.467352                               

            Emisión primaria y multiplicador (millones S/) - Circulante Desestacionalizado  \
Fecha                                                                                        
2003-09-01                                           8.591877                                
2003-10-01                                           8.602051                                
2003-11-01                                           8.613430                                
2003-12-01                                           8.629901                                
2004-01-01                                           8.649204                                

            Liquidez internacional del BCRP - RIN - Reservas Internacionales Netas (millones US$)  \
Fecha                                                                                               
2003-09-01                                           9.185581                                       
2003-10-01                                           9.191108                                       
2003-11-01                                           9.240173                                       
2003-12-01                                           9.229584                                       
2004-01-01                                           9.265177                                       

            Remuneraciones - Remuneración Mínima Vital - Índice Real (base 1994 = 100)  \
Fecha                                                                                    
2003-09-01                                           5.440117                            
2003-10-01                                           5.495508                            
2003-11-01                                           5.493832                            
2003-12-01                                           5.488208                            
2004-01-01                                           5.482848                            

            Cotizaciones de productos (promedio del periodo) - Trigo - EEUU (US$ por tonelada

In [49]:
df.tail()

Índice de precios Lima Metropolitana (var% mensual) - IPC  \
Fecha                                                                   
2023-08-01                                           0.182324           
2023-09-01                                           0.133933           
2023-10-01                                           0.085365           
2023-11-01                                           0.036736           
2023-12-01                                          -0.011845           

            Tasas de interés del Banco Central de Reserva  - Tasa de Referencia de la Política Monetaria  \
Fecha                                                                                                      
2023-08-01                                               7.75                                              
2023-09-01                                               7.50                                              
2023-10-01                                               7.25                                              
2023-11-01                                               7.00                                              
2023-12-01                                               6.75                                              

            Tipo de cambio - promedio del periodo (S/ por US$) - Interbancario - Promedio  \
Fecha                                                                                       
2023-08-01                                           3.697768                               
2023-09-01                                           3.730995                               
2023-10-01                                           3.845759                               
2023-11-01                                           3.760795                               
2023-12-01                                           3.733942                               

            Emisión primaria y multiplicador (millones S/) - Circulante Desestacionalizado  \
Fecha                                                                                        
2023-08-01                                          11.194037                                
2023-09-01                                          11.196305                                
2023-10-01                                          11.202967                                
2023-11-01                                          11.200187                                
2023-12-01                                          11.186627                                

            Liquidez internacional del BCRP - RIN - Reservas Internacionales Netas (millones US$)  \
Fecha                                                                                               
2023-08-01                                          11.182374                                       
2023-09-01                                          11.173721                                       
2023-10-01                                          11.172515                                       
2023-11-01                                          11.180961                                       
2023-12-01                                          11.170900                                       

            Remuneraciones - Remuneración Mínima Vital - Índice Real (base 1994 = 100)  \
Fecha                                                                                    
2023-08-01                                           5.646268                            
2023-09-01                                           5.646104                            
2023-10-01                                           5.649334                            
2023-11-01                                           5.650966                            
2023-12-01                                           5.646966                            

            Cotizaciones de productos (promedio del periodo) - Trigo - EEUU (US$ por tonelada

## 1.3 Data Inspection
We inspect the df. We first verify that all values are non-null. The, we apply the ```describe``` function to see the main variables.

In [50]:
df.isna().sum()

Índice de precios Lima Metropolitana (var% mensual) - IPC                                       0
Tasas de interés del Banco Central de Reserva  - Tasa de Referencia de la Política Monetaria    0
Tipo de cambio - promedio del periodo (S/ por US$) - Interbancario - Promedio                   0
Emisión primaria y multiplicador (millones S/) - Circulante Desestacionalizado                  0
Liquidez internacional del BCRP - RIN - Reservas Internacionales Netas (millones US$)           0
Remuneraciones - Remuneración Mínima Vital - Índice Real (base 1994 = 100)                      0
Cotizaciones de productos (promedio del periodo) - Trigo - EEUU (US$ por toneladas)             0
Cotizaciones de productos (promedio del periodo) - Maíz - EEUU (US$ por toneladas)              0
Cotizaciones de productos (promedio del periodo) - Aceite Soya - EEUU (US$ por toneladas)       0
Cotizaciones de productos (promedio del periodo) - Petróleo - WTI (US$ por barriles)            0
dtype: int64

In [51]:
df.describe()

Índice de precios Lima Metropolitana (var% mensual) - IPC  \
count                                         244.000000           
mean                                            0.268869           
std                                             0.148128           
min                                            -0.011845           
25%                                             0.157445           
50%                                             0.248044           
75%                                             0.330206           
max                                             0.679208           

       Tasas de interés del Banco Central de Reserva  - Tasa de Referencia de la Política Monetaria  \
count                                         244.000000                                              
mean                                            3.658811                                              
std                                             1.710424                                              
min                                             0.250000                                              
25%                                             2.750000                                              
50%                                             3.750000                                              
75%                                             4.250000                                              
max                                             7.750000                                              

       Tipo de cambio - promedio del periodo (S/ por US$) - Interbancario - Promedio  \
count                                         244.000000                               
mean                                            3.214740                               
std                                             0.363194                               
min                                             2.552173                               
25%                                             2.870174                               
50%                                             3.251714                               
75%                                             3.399687                               
max                                             4.108055                               

       Emisión primaria y multiplicador (millones S/) - Circulante Desestacionalizado  \
count                                         244.000000                                
mean                                           10.207171                                
std                                             0.766773                                
min                                             8.591877                                
25%                                             9.678975                                
50%                                            10.392176                                
75%                                            10.748751                                
max                                            11.305032                                

       Liquidez internacional del BCRP - RIN - Reservas Internacionales Netas (millones US$)  \
count                                         244.000000                                       
mean                                           10.662819                                       
std                                             0.625920                                       
min                                             9.185581                                       
25%                                            10.365558                                       
50%                                            11.010612                                       
75%                                            11.107878                                       
max                                            11.288803                                       

      

We have 195 observation ranging from ```2003-10-01``` to ```2019-12-01```. The mean of monthly % change of all CPI variables is around 0.2. The mean of the lacing rate and the interest rate is 10.7% and 3.67%, respectively. The three monetary variables have a small monthly % change, around 0.01 and 0.001 for Minimum Wage index. 

## 1.4 Data adjustment
We will rename the columns for easier identification of the variables. We will also create a new dataframe with the lags of the variables. 

In [52]:
# New column names
columns = {
    'Índice de precios Lima Metropolitana (var% mensual) - IPC': 'CPI',
    'Tasas de interés del Banco Central de Reserva  - Tasa de Referencia de la Política Monetaria': 'Monetary Policy Rate',
    'Tipo de cambio - promedio del periodo (S/ por US$) - Interbancario - Promedio' : 'Exchange rate', 
    'Emisión primaria y multiplicador (millones S/) - Circulante Desestacionalizado': 'Circulating Currency Seasonally Adjusted (mill S/)',
    'Liquidez internacional del BCRP - RIN - Reservas Internacionales Netas (millones US$)': 'Net International Reserves (mill $)',
    'Remuneraciones - Remuneración Mínima Vital - Índice Real (base 1994 = 100)': 'Real Minimum Wage (Index)',
    'Cotizaciones de productos (promedio del periodo) - Trigo - EEUU (US$ por toneladas)': 'Wheat (US$ per ton)',
    'Cotizaciones de productos (promedio del periodo) - Maíz - EEUU (US$ por toneladas)': 'Corn  (US$ per ton)',
    'Cotizaciones de productos (promedio del periodo) - Aceite Soya - EEUU (US$ por toneladas)': 'Soybean oil (US$ per ton)',
    'Cotizaciones de productos (promedio del periodo) - Petróleo - WTI (US$ por barriles)': 'Crude oil (US$ per barrel)'  
}

# We rename the columns so they are easier to analyse
df.rename(columns=columns, inplace=True)

In [53]:
df_lags = df.copy()

for variable in df_lags.columns[1:]:
    df_lags[f'{variable}_lag_1'] = df_lags[variable].shift()
    df_lags[f'{variable}_lag_2'] = df_lags[variable].shift(2)
    df_lags[f'{variable}_lag_3'] = df_lags[variable].shift(3)
    df_lags[f'{variable}_lag_4'] = df_lags[variable].shift(4)

In [54]:
# We delete contemporary variables
df_lags.drop(columns = ['Monetary Policy Rate','Circulating Currency Seasonally Adjusted (mill S/)',
       'Net International Reserves (mill $)', 'Real Minimum Wage (Index)', 'Wheat (US$ per ton)', 'Corn  (US$ per ton)', 
       'Soybean oil (US$ per ton)', 'Crude oil (US$ per barrel)', 'Exchange rate'], inplace = True)

df_lags = df_lags.dropna()

## 1.5 Save Results
We save it to the ```input``` folder, where we can use it to do the forecasting in the next notebook.

In [55]:
df.to_csv('../../../input/df_raw_test.csv')

In [56]:
df_lags.to_csv('../../../input/df_lags_test.csv')

In [57]:
df_lags.tail()

CPI  Monetary Policy Rate_lag_1  Monetary Policy Rate_lag_2  \
Fecha                                                                          
2023-08-01  0.182324                        7.75                        7.75   
2023-09-01  0.133933                        7.75                        7.75   
2023-10-01  0.085365                        7.50                        7.75   
2023-11-01  0.036736                        7.25                        7.50   
2023-12-01 -0.011845                        7.00                        7.25   

            Monetary Policy Rate_lag_3  Monetary Policy Rate_lag_4  \
Fecha                                                                
2023-08-01                        7.75                        7.75   
2023-09-01                        7.75                        7.75   
2023-10-01                        7.75                        7.75   
2023-11-01                        7.75                        7.75   
2023-12-01                        7.50                        7.75   

            Exchange rate_lag_1  Exchange rate_lag_2  Exchange rate_lag_3  \
Fecha                                                                       
2023-08-01             3.601255             3.650419             3.688668   
2023-09-01             3.697768             3.601255             3.650419   
2023-10-01             3.730995             3.697768             3.601255   
2023-11-01             3.845759             3.730995             3.697768   
2023-12-01             3.760795             3.845759             3.730995   

            Exchange rate_lag_4  \
Fecha                             
2023-08-01             3.765450   
2023-09-01             3.688668   
2023-10-01             3.650419   
2023-11-01             3.601255   
2023-12-01             3.697768   

            Circulating Currency Seasonally Adjusted (mill S/)_lag_1  ...  \
Fecha                                                                 ...   
2023-08-01                                          11.196625         ...   
2023-09-01                                          11.194037         ...   
2023-10-01                                          11.196305         ...   
2023-11-01                                          11.202967         ...   
2023-12-01                                          11.200187         ...   

            Corn  (US$ per ton)_lag_3  Corn  (US$ per ton)_lag_4  \
Fecha                                                              
2023-08-01                   5.461808                   5.531248   
2023-09-01                   5.487035                   5.461808   
2023-10-01                   5.368204                   5.487035   
2023-11-01                   5.288844                   5.368204   
2023-12-01                   5.178940                   5.288844   

            Soybean oil (US$ per ton)_lag_1  Soybean oil (US$ per ton)_lag_2  \
Fecha                                                                          
2023-08-01                         7.342273                         7.158705   
2023-09-01                         7.354495                         7.342273   
2023-10-01                         7.283402                         7.354495   
2023-11-01                         7.138916                         7.283402   
2023-12-01                         7.076030                         7.138916   

            Soybean oil (US$ per ton)_lag_3  Soybean oil (US$ per ton)_lag_4  \
Fecha                                                                          
2023-08-01                         7.063081                         7.124298   
2023-09-01                         7.158705                         7.063081   
2023-10-01                         7.342273                         7.158705   
2023-11-01                         7.354495                         7.342273   
2023-12-01                         7.283402                         7.354495   

            Crude oil (US$ per barrel)_l

In [58]:
df.tail()

CPI  Monetary Policy Rate  Exchange rate  \
Fecha                                                       
2023-08-01  0.182324                  7.75       3.697768   
2023-09-01  0.133933                  7.50       3.730995   
2023-10-01  0.085365                  7.25       3.845759   
2023-11-01  0.036736                  7.00       3.760795   
2023-12-01 -0.011845                  6.75       3.733942   

            Circulating Currency Seasonally Adjusted (mill S/)  \
Fecha                                                            
2023-08-01                                          11.194037    
2023-09-01                                          11.196305    
2023-10-01                                          11.202967    
2023-11-01                                          11.200187    
2023-12-01                                          11.186627    

            Net International Reserves (mill $)  Real Minimum Wage (Index)  \
Fecha                                                                        
2023-08-01                            11.182374                   5.646268   
2023-09-01                            11.173721                   5.646104   
2023-10-01                            11.172515                   5.649334   
2023-11-01                            11.180961                   5.650966   
2023-12-01                            11.170900                   5.646966   

            Wheat (US$ per ton)  Corn  (US$ per ton)  \
Fecha                                                  
2023-08-01             5.714305             5.288844   
2023-09-01             5.677783             5.178940   
2023-10-01             5.548594             5.191875   
2023-11-01             5.567082             5.146240   
2023-12-01             5.599153             5.145342   

            Soybean oil (US$ per ton)  Crude oil (US$ per barrel)  
Fecha                                                              
2023-08-01                   7.354495                    4.399204  
2023-09-01                   7.283402                    4.490772  
2023-10-01                   7.138916                    4.448655  
2023-11-01                   7.076030                    4.351357  
2023-12-01                   7.046320                    4.276196

In [59]:
df_lags.columns

Index(['CPI', 'Monetary Policy Rate_lag_1', 'Monetary Policy Rate_lag_2',
       'Monetary Policy Rate_lag_3', 'Monetary Policy Rate_lag_4',
       'Exchange rate_lag_1', 'Exchange rate_lag_2', 'Exchange rate_lag_3',
       'Exchange rate_lag_4',
       'Circulating Currency Seasonally Adjusted (mill S/)_lag_1',
       'Circulating Currency Seasonally Adjusted (mill S/)_lag_2',
       'Circulating Currency Seasonally Adjusted (mill S/)_lag_3',
       'Circulating Currency Seasonally Adjusted (mill S/)_lag_4',
       'Net International Reserves (mill $)_lag_1',
       'Net International Reserves (mill $)_lag_2',
       'Net International Reserves (mill $)_lag_3',
       'Net International Reserves (mill $)_lag_4',
       'Real Minimum Wage (Index)_lag_1', 'Real Minimum Wage (Index)_lag_2',
       'Real Minimum Wage (Index)_lag_3', 'Real Minimum Wage (Index)_lag_4',
       'Wheat (US$ per ton)_lag_1', 'Wheat (US$ per ton)_lag_2',
       'Wheat (US$ per ton)_lag_3', 'Wheat (US$ per ton)_l